acceptancetests/assess_upgrade.py

```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python
+""" Assess upgrading juju controllers and models.
+
+Bootstrap a previous version of juju and then upgrade the controller and models
+to the new one.
+
+  - Spins up a local streams server (requires to be run on lxd or on a machine
+     with an externally routable interface.
+  - Bootstraps using the 'stable' juju then upgrades to the 'devel' version.
+  - Deploys mediawiki and mysql to ensure workloads continue working.
+"""
+
+from __future__ import print_function
+
+import argparse
+import logging
+import sys
+from subprocess import (
+    CalledProcessError,
+)
+
+
+from collections import (
+    namedtuple,
+    )
+from utility import (
+    add_basic_testing_arguments,
+    configure_logging,
+    temp_dir,
+    )
+from jujupy.binaries import (
+    get_stable_juju
+    )
+from jujupy.client import (
+    ModelClient,
+    get_stripped_version_number,
+)
+from jujupy.stream_server import (
+    StreamServer,
+    agent_tgz_from_juju_binary,
+    )
+from jujupy.wait_condition import (
+    ModelCheckFailed,
+    WaitModelVersion,
+    wait_for_model_check,
+)
+from jujupy.workloads import (
+    deploy_mediawiki_with_db,
+    assert_media_wiki_is_responding
+    )
+
+from utility import (
+    JujuAssertionError
+)
+
+__metaclass__ = type
+
+
+log = logging.getLogger("assess_upgrade")
+
+
+VersionParts = namedtuple('VersionParts', ['version', 'series', 'arch'])
+
+
+def assess_upgrade_from_stable_to_develop(args, stable_bsm, devel_client):
+    stable_client = stable_bsm.client
+    with temp_dir() as base_dir:
+        stream_server = StreamServer(base_dir)
+        setup_agent_metadata(
+            stream_server, args.stable_agent,
+            stable_client, base_dir, 'proposed')
+        setup_agent_metadata(
+            stream_server, args.devel_agent,
+            devel_client, base_dir, 'proposed')
+        with stream_server.server() as url:
+            stable_client.env.update_config({
+                'agent-metadata-url': url,
+                'agent-stream': 'proposed'
+            })
+            with stable_bsm.booted_context(False):
+                assert_stable_model_is_correct(stable_client)
```

Comment:
everything in this method, including this line I assume would be common to other scenarios right?